content/4_Gold/Combinatorics.mdx

```diff
@@ -807,6 +807,102 @@ print()
 </PySection>
 </LanguageSection>
 
+## Catalan Numbers
+
+The [Catalan numbers](https://en.wikipedia.org/wiki/Catalan_number) are a sequence of positive integers that can be very useful in counting problems in combinatorics.
+The $n$-th Catalan can be expressed as following using binomial coeffiecients:
+
+$$
+C_n=\frac{1}{n+1}\cdot \binom{2n}{n}=\frac{(2n)!}{(n+1)!\,n!}
+$$
+
+They also have the recurrence formula
+$$
+C_n= \sum^{n}_{i=1}{C_i \cdot C_{n-i}} \,\,\, \text{ for } n>0 \\
+$$
+which can also be expressed as
+$$
+C_n=\frac{2(2n-1)}{n+1} \cdot C_{n-1}
+$$
+
+The first $5$ Catalan numbers are
+| n | 0| 1 | 2 | 3 | 4 | 5 |
+|---| -| - | - | - | - | - |
+| $C_n$| 1 | 1 | 2 | 5 | 14| 42 |
+
+### Applications
+
+The Catalan numbers can be used to represent a wide variety of things.
+
+For example, $C_n$ is equal to the number of valid parenthesis expressions of length $2n$.
+Take, for instance, $C_3=5$ a:
+* `()()()`
+* `(())()`
+* `()(())`
+* `((()))`
+* `(()())`
+
+It's also equal to the number of full binary trees with $n+1$ leaves.
+The following image shows the $5$ binary trees with $4$ leaves:
+
+![binary-trees](./assets/catalan-binary-trees.png)
+
+$C_n$ is also the number of monotonic lattice paths along the edges of a $n \times n$ grid that don't pass above the diagonal.
+The paths start in the lower left corner and ends in the upper right corner.
+
+For example, there are $C_4=14$ paths in a $4 \times 4 grid$:
+
+![lattice points](./assets/catalan-lattice-paths.png)
+
+The next two examples are a bit more niche, but they're still interesting to think about.
+
+Consider a convex polygon with $n+2$ sides divided into $n$ triangles by connecting vertices with non-intersecting lines.
+The number of different ways to divide the polygon in this way is equal to $C_n$.
+
+Here's the particular case for $n=3$ in which we have $C_4=$:
+
+![polygons](./assets/catalan-polygons.png)
+
+$C_n$ is also equal to the number of mountain ranges - see [Dyck words](https://en.wikipedia.org/wiki/Dyck_language) - of length $2n$ consisting of $n$ upstrokes and $n$ downstrokes.
+
+![mountains](./assets/catalan-mountain-strokes.png)
+
+<FocusProblem problem="sample3" />
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+
+using namespace std;
+
+const int MOD = 1e6;
+const int MAX_N = 2e3;
+
+int main() {
+	int comb[MAX_N][MAX_N];
```

Comment:
you'd think a better name would be `catalan`

Reply:
done